### PR TITLE
feat(Core/Hooks): Functions to improve core-based modding support

### PR DESCRIPTION
**Pull request by gitdnd**
Added two new hooks, BeforeCastTime that runs after a spell is prepared but before begging it's cast time and WhileCast that runs every time the update function for a spell runs while casting or channeling, as long as m_timer does not become 0. 

Added ModifySpellValue for more easily modifying spell values more easily during more awkward points in a spells lifetime.

Added an std::map<MapDummy, std::optional<std::any>> triggerDummy; where MapDummy is a class enum, to allow modded spells to have optional data assigned to them for more complex spells, especially ones that happen from interactions between different spells, such as using one spell to cast another, useful for spells that rely on one or more dummy spells that use more complex input from one "phase" to another.

## Changes Proposed:
-  Add two new hooks, BeforeCastTime and WhileCast.
-  Added function ModifySpellValue 
- Added std::map<MapDummy, std::optional<std::any>> to easily track data from one spell to another.

## Issues Addressed:
None

## SOURCE:
I made it up

## Tests Performed:
- Builds and runs without errors.
- Example of spell that utilizes new hooks to tell if a player is in the air while casting the spell, delaying the cast time until they land and gaining damage based on how long they were in the air. https://streamable.com/5q7arz

## How to Test the Changes:
1. Create a spell script that on each event of the new hooks casts a spell with triggered = true and cast it.
2. Create a spell script that modifies effect values on hook using ModifySpellValue and cast it.
3. Create a spell called BetterPyroblast that casts Pyroblast using hooks and use FindCurrentSpellBySpellId to find the Pyroblast, and an enum in MapDummy called CastByBetterPyroblast. Use GetTriggerDummy to get the reference to triggerDummy and set triggerDummy[MapDummy::CastByBetterPyroblast] = true. Create an on-hit hook for Pyroblast that detects if triggerDummy[MapDummy::CastByBetterPyroblast] has a value and check if it's equal to true, casts a third dummy stun on the target.

## Known Issues and TODO List:
- None

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Review by Kitzunu — CHANGES_REQUESTED** [review]
Dont change codestyle

When use`enum class` please assign an underlying type and write the enums like put together words like 

```cpp
enum class Classes : uint8
{
    SuperWarrior,
    Warrior
    UnbeatableCritterClass
}

File: src/server/game/Spells/Spell.h [hunk 1]
```diff
@@ -559,14 +570,16 @@ class Spell
...
     void SetSpellInfo(SpellInfo const* info) { m_spellInfo = info; }
     int32 GetPowerCost() const { return m_powerCost; }
 
     bool UpdatePointers();                              // must be used at call Spell code after time delay (non triggered spell cast/update spell call/etc)
 
     void CleanupTargetList();
 
-    void SetSpellValue(SpellValueMod mod, int32 value);
+    void              SetSpellValue(SpellValueMod mod, int32 value);
+    void              ModifySpellValue(SpellValueMod mod, int32 value);
```
Comment (Nefertumm) [inline src/server/game/Spells/Spell.h:582]: fix indent here too please.

File: src/server/game/Spells/Spell.h [hunk 2]
```diff
@@ -706,7 +719,9 @@ class Spell
     // Scripting system
     bool _scriptsLoaded;
     //void LoadScripts();
-    void CallScriptBeforeCastHandlers();
+    void                    CallScriptBeforeCastTimeHandlers();
+    void                    CallScriptBeforeCastHandlers();
+    void                    CallScriptWhileCastHandlers();
```
Comment (Nefertumm) [inline src/server/game/Spells/Spell.h:724]: here as well, is this clang indent? :P

File: src/server/game/Spells/Spell.cpp [hunk 3]
```diff
@@ -7915,30 +7921,60 @@ void Spell::SetSpellValue(SpellValueMod mod, int32 value)
...
-        case SPELLVALUE_AURA_STACK:
-            m_spellValue->AuraStackAmount = uint8(value);
-            break;
-        case SPELLVALUE_AURA_DURATION:
-            m_spellValue->AuraDuration = value;
-            break;
-        case SPELLVALUE_FORCED_CRIT_RESULT:
-            m_spellValue->ForcedCritResult = (bool)value;
-            break;
+    case SPELLVALUE_BASE_POINT0:
```
Comment (Nefertumm) [inline src/server/game/Spells/Spell.cpp:7924]: indentantion got messy

File: src/server/game/Spells/Spell.cpp [hunk 4]
```diff
@@ -7915,30 +7921,60 @@ void Spell::SetSpellValue(SpellValueMod mod, int32 value)
...
+        m_spellValue->AuraDuration = value;
+        break;
+    case SPELLVALUE_FORCED_CRIT_RESULT:
+        m_spellValue->ForcedCritResult = (bool) value;
+        break;
+    }
+}
+void Spell::ModifySpellValue(SpellValueMod mod, int32 value)
+{
+    switch (mod)
```
Comment (Nefertumm) [inline src/server/game/Spells/Spell.cpp:7952]: here too

**Comment by Nefertumm** [comment]
Thank you for your PR! 💯

**Review by Nefertumm — APPROVED** [review]
Seems ok now! Also, when we add new hooks, we should add a changelog to keep track of new additions.
Read this one! https://www.azerothcore.org/wiki/how-to-use-changelog
Thank you!

File: src/server/game/Spells/Spell.cpp [hunk 5]
```diff
@@ -7996,15 +8032,36 @@ void Spell::LoadScripts()
         ++itr;
     }
 }
+void Spell::CallScriptBeforeCastTimeHandlers()
+{
+    for (std::list<SpellScript*>::iterator scritr = m_loadedScripts.begin(); scritr != m_loadedScripts.end(); ++scritr)
+    {
+        (*scritr)->_PrepareScriptCall(SPELL_SCRIPT_HOOK_BEFORE_CAST_TIME);
+        std::list<SpellScript::CastHandler>::iterator hookItrEnd = (*scritr)->BeforeCastTime.end(), hookItr = (*scritr)->BeforeCastTime.begin();
+        for (; hookItr != hookItrEnd; ++hookItr) (*hookItr).Call(*scritr);
```
Comment (Nefertumm) [inline src/server/game/Spells/Spell.cpp:8041]:
```suggestion
        for (; hookItr != hookItrEnd; ++hookItr)
        {
            (*hookItr).Call(*scritr);
        }
```

File: src/server/game/Spells/Spell.cpp [hunk 6]
```diff
@@ -7996,15 +8032,36 @@ void Spell::LoadScripts()
...
+        (*scritr)->_FinishScriptCall();
+    }
+}
+void Spell::CallScriptWhileCastHandlers()
+{
+    for (std::list<SpellScript*>::iterator scritr = m_loadedScripts.begin(); scritr != m_loadedScripts.end(); ++scritr)
+    {
+        (*scritr)->_PrepareScriptCall(SPELL_SCRIPT_HOOK_WHILE_CAST);
+        std::list<SpellScript::CastHandler>::iterator hookItrEnd = (*scritr)->WhileCast.end(), hookItr = (*scritr)->WhileCast.begin();
+        for (; hookItr != hookItrEnd; ++hookItr) (*hookItr).Call(*scritr);
```
Comment (Nefertumm) [inline src/server/game/Spells/Spell.cpp:8064]:
```suggestion
        for (; hookItr != hookItrEnd; ++hookItr)
        {
            (*hookItr).Call(*scritr);
        }
```

**Review by Nefertumm — COMMENTED** [review]
I almost missed those too

**Comment by gitdnd** [comment]
https://github.com/azerothcore/wiki/pull/680

> Thank you for your PR! 💯 EDIT: Oh yeah, I almost forgot, you will need to add wiki related stuff :)

Like this?

**Comment by acidmanifesto** [comment]
updated title of PR to reflect pr title standards

File: src/server/game/Spells/Spell.cpp [hunk 7]
```diff
@@ -7937,7 +7943,37 @@ void Spell::SetSpellValue(SpellValueMod mod, int32 value)
             m_spellValue->AuraDuration = value;
             break;
         case SPELLVALUE_FORCED_CRIT_RESULT:
-            m_spellValue->ForcedCritResult = (bool)value;
+            m_spellValue->ForcedCritResult = (bool) value;
```
Comment (Kitzunu) [inline src/server/game/Spells/Spell.cpp:7946]: revert these changes, everywhere

**Review by Kitzunu — CHANGES_REQUESTED** [review]

File: src/server/game/Spells/Spell.cpp [hunk 8]
```diff
@@ -7996,15 +8032,42 @@ void Spell::LoadScripts()
...
         (*scritr)->_PrepareScriptCall(SPELL_SCRIPT_HOOK_BEFORE_CAST);
         std::list<SpellScript::CastHandler>::iterator hookItrEnd = (*scritr)->BeforeCast.end(), hookItr = (*scritr)->BeforeCast.begin();
+        for (; hookItr != hookItrEnd; ++hookItr) (*hookItr).Call(*scritr);
+
+        (*scritr)->_FinishScriptCall();
+    }
+}
+void Spell::CallScriptWhileCastHandlers()
+{
+    for (std::list<SpellScript*>::iterator scritr = m_loadedScripts.begin(); scritr != m_loadedScripts.end(); ++scritr)
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/Spell.cpp:8063]: for (auto const& itr : m_loadedScripts)

Comment (gitdnd) [reply src/server/game/Spells/Spell.cpp:8063]: Wouldn't that be inconsistent with how the rest of the hooks do their for loop?

File: src/server/game/Spells/Spell.cpp [hunk 9]
```diff
@@ -7941,6 +7947,36 @@ void Spell::SetSpellValue(SpellValueMod mod, int32 value)
             break;
     }
 }
+void Spell::ModifySpellValue(SpellValueMod mod, int32 value)
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/Spell.cpp:7950]:
```suggestion

void Spell::ModifySpellValue(SpellValueMod mod, int32 value)
```

File: src/server/game/Spells/Spell.cpp [hunk 10]
```diff
@@ -7996,15 +8032,42 @@ void Spell::LoadScripts()
         ++itr;
     }
 }
+void Spell::CallScriptBeforeCastTimeHandlers()
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/Spell.cpp:8035]:
```suggestion

void Spell::CallScriptBeforeCastTimeHandlers()
```

File: src/server/game/Spells/Spell.cpp [hunk 11]
```diff
@@ -7996,15 +8032,42 @@ void Spell::LoadScripts()
...
+{
+    for (std::list<SpellScript*>::iterator scritr = m_loadedScripts.begin(); scritr != m_loadedScripts.end(); ++scritr)
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/Spell.cpp:8037]: https://en.cppreference.com/w/cpp/language/range-for

File: src/server/game/Spells/Spell.cpp (hunk 8 above)
Comment (Winfidonarleyan) [reply src/server/game/Spells/Spell.cpp:8063]: example?

File: src/server/game/Spells/Spell.h [hunk 12]
```diff
@@ -18,6 +18,10 @@
 #ifndef __SPELL_H
 #define __SPELL_H
 
+#include <any>
+#include <map>
+#include <optional>
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/Spell.h:23]: after all includes

Comment (Winfidonarleyan) [inline src/server/game/Spells/Spell.h:23]: "Optional.h"

File: src/server/game/Spells/Spell.h [hunk 13]
```diff
@@ -767,6 +782,30 @@ class Spell
...
+    /// A collection of random stuff to be read by a single spell.
+    /// includes "which spell issued the cast of this spell" so I can tell from the finishing of basic attack if it was cast by undead strike or crusader strike etc.
+    /// </summary>
+
+public:
+    int32 GetSpellTimer()
+    {
+        return m_timer;
+    };
+    void ModifySpellTimer(int32 amount)
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/Spell.h:796]:
```suggestion

    void ModifySpellTimer(int32 amount)
```

File: src/server/game/Spells/Spell.h [hunk 14]
```diff
@@ -767,6 +782,30 @@ class Spell
...
+    {
+        m_timer += amount;
+    };
+    void SetSpellTimer(int32 amount)
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/Spell.h:800]:
```suggestion

    void SetSpellTimer(int32 amount)
```

File: src/server/game/Spells/Spell.h [hunk 15]
```diff
@@ -767,6 +782,30 @@ class Spell
...
+    {
+        m_timer = amount;
+    };
+    std::map<MapDummy, std::optional<std::any>> triggerDummy = {};
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/Spell.h:804]:
```suggestion

    std::map<MapDummy, std::optional<std::any>> triggerDummy = {};
```

File: src/server/game/Spells/Spell.cpp (hunk 8 above)
Comment (gitdnd) [reply src/server/game/Spells/Spell.cpp:8065]:
```

void Spell::CallScriptAfterCastHandlers()
{
    for (std::list<SpellScript*>::iterator scritr = m_loadedScripts.begin(); scritr != m_loadedScripts.end(); ++scritr)
    {
        (*scritr)->_PrepareScriptCall(SPELL_SCRIPT_HOOK_AFTER_CAST);
        std::list<SpellScript::CastHandler>::iterator hookItrEnd = (*scritr)->AfterCast.end(), hookItr = (*scritr)->AfterCast.begin();
        for (; hookItr != hookItrEnd; ++hookItr)
            (*hookItr).Call(*scritr);

        (*scritr)->_FinishScriptCall();
    }
}

```
Is how I see it's done elsewhere in the code.

File: src/server/game/Spells/Spell.h [hunk 16]
```diff
@@ -18,12 +18,17 @@
 #ifndef __SPELL_H
 #define __SPELL_H
 
+
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/Spell.h:21]:
```suggestion

```

File: src/server/game/Spells/Spell.h [hunk 17]
```diff
@@ -18,12 +18,17 @@
...
 #include "GridDefines.h"
 #include "ObjectMgr.h"
 #include "PathGenerator.h"
 #include "SharedDefines.h"
 #include "SpellInfo.h"
 
+#include <any>
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/Spell.h:28]:
```suggestion
#include "SpellInfo.h"
#include <any>
```

File: src/server/game/Spells/Spell.cpp [hunk 18]
```diff
@@ -7997,14 +8034,42 @@ void Spell::LoadScripts()
...
     for (std::list<SpellScript*>::iterator scritr = m_loadedScripts.begin(); scritr != m_loadedScripts.end(); ++scritr)
     {
         (*scritr)->_PrepareScriptCall(SPELL_SCRIPT_HOOK_BEFORE_CAST);
         std::list<SpellScript::CastHandler>::iterator hookItrEnd = (*scritr)->BeforeCast.end(), hookItr = (*scritr)->BeforeCast.begin();
+        for (; hookItr != hookItrEnd; ++hookItr) (*hookItr).Call(*scritr);
+
+        (*scritr)->_FinishScriptCall();
+    }
+}
+void Spell::CallScriptWhileCastHandlers()
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/Spell.cpp:8063]:
```suggestion

void Spell::CallScriptWhileCastHandlers()
```

File: src/server/game/Spells/Spell.cpp [hunk 19]
```diff
@@ -7997,14 +8034,42 @@ void Spell::LoadScripts()
...
+    }
+}
+
 void Spell::CallScriptBeforeCastHandlers()
 {
     for (std::list<SpellScript*>::iterator scritr = m_loadedScripts.begin(); scritr != m_loadedScripts.end(); ++scritr)
     {
         (*scritr)->_PrepareScriptCall(SPELL_SCRIPT_HOOK_BEFORE_CAST);
         std::list<SpellScript::CastHandler>::iterator hookItrEnd = (*scritr)->BeforeCast.end(), hookItr = (*scritr)->BeforeCast.begin();
+        for (; hookItr != hookItrEnd; ++hookItr) (*hookItr).Call(*scritr);
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/Spell.cpp:8058]: brackets

File: src/server/game/Spells/Spell.h [hunk 20]
```diff
@@ -17,12 +17,14 @@
 
 #ifndef __SPELL_H
 #define __SPELL_H
-
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/Spell.h:20]: no need delete this line

File: src/server/game/Spells/Spell.h [hunk 21]
```diff
@@ -767,6 +780,33 @@ class Spell
...
+    /// <summary>
+    /// A collection of random stuff to be read by a single spell.
+    /// includes "which spell issued the cast of this spell" so I can tell from the finishing of basic attack if it was cast by undead strike or crusader strike etc.
+    /// </summary>
+
+public:
+    int32 GetSpellTimer()
+    {
+        return m_timer;
+    };
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/Spell.h:793]:
```suggestion
    }
```

File: src/server/game/Spells/Spell.h [hunk 22]
```diff
@@ -767,6 +780,33 @@ class Spell
...
+
+    void ModifySpellTimer(int32 amount)
+    {
+        m_timer += amount;
+    };
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/Spell.h:798]:
```suggestion
    }
```

File: src/server/game/Spells/Spell.h [hunk 23]
```diff
@@ -767,6 +780,33 @@ class Spell
...
+
+    void SetSpellTimer(int32 amount)
+    {
+        m_timer = amount;
+    };
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/Spell.h:803]:
```suggestion
    }
```

File: src/server/game/Spells/Spell.h [hunk 24]
```diff
@@ -767,6 +780,33 @@ class Spell
...
+
+    std::map<MapDummy, std::optional<std::any>> triggerDummy = {};
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/Spell.h:805]:
```suggestion
    std::map<MapDummy, Optional<std::any>> triggerDummy = {};
```

File: src/server/game/Spells/Spell.h [hunk 25]
```diff
@@ -767,6 +780,33 @@ class Spell
...
+    std::map<MapDummy, std::optional<std::any>>& GetTriggerDummy()
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/Spell.h:806]:
```suggestion
    std::map<MapDummy, Optional<std::any>>& GetTriggerDummy()
```

**Comment by Kitzunu** [comment]
Status?
